src/js/controller/controller.js

```diff
@@ -280,33 +274,119 @@ define([
                     var entry = entries[0];
                     if (entry.target === _this.getContainer()) {
                         _model.set('intersectionRatio', entry.intersectionRatio);
-                        _checkVisibility();
                     }
                 }
             }
 
-            function _onActiveTabChange() {
-                _checkVisibility();
-            }
+            function _getVisibility() {
+                // Set visibility to 1 if we're in fullscreen
+                var fullscreen = _model.get('fullscreen');
+                if (fullscreen) {
+                    return 1;
+                }
 
-            function _checkVisibility() {
                 // Set visibility to 0 if we're not in the active tab
+                var activeTab = _model.get('activeTab');
+                if (!activeTab) {
+                    return 0;
+                }
                 // Otherwise, set it to the intersection ratio reported from the intersection observer
                 var intersectionRatio = _model.get('intersectionRatio');
-                var activeTab = _model.get('activeTab');
 
-                // We don't yet know if the player is viewable
-                if (activeTab && _.isUndefined(intersectionRatio)) {
-                    return;
+                if (_.isUndefined(intersectionRatio)) {
+                    // Set visibility to 1 if we're in an iFrame and intersection is unknown
+                    var iFrame = _model.get('iFrame');
+                    if (iFrame) {
+                        return 1;
+                    }
+
+                    // Get intersectionRatio through brute force
+                    intersectionRatio = _computeVisibility(_view.element());
                 }
 
-                var visibility = activeTab ? intersectionRatio : 0;
-                _model.set('visibility', visibility);
+                return intersectionRatio;
+            }
+
+            function _computeVisibility(target) {
```

Comment:
How hard would it be to put this function in a util so we avoid +90 LOC in the controller?